src/main/scala/zio/sqs/SqsPublisher.scala

```diff
@@ -34,4 +38,89 @@ object SqsPublisher {
     }
+
+  trait Event {
+    def body: String
+    def attributes: Map[String, MessageAttributeValue]
+    def groupId: Option[String]
+    def deduplicationId: Option[String]
+  }
+
+  final case class StringEvent(body: String) extends Event {
+    override def attributes: Map[String, MessageAttributeValue] = Map.empty[String, MessageAttributeValue]
+    override def groupId: Option[String]                        = None
+    override def deduplicationId: Option[String]                = None
+  }
+
+  type ErrorOrEvent = Either[(BatchResultErrorEntry, Event), Event]
+
+  def stringStream(
+    client: SqsAsyncClient,
+    queueUrl: String,
+    settings: SqsPublisherStreamSettings = SqsPublisherStreamSettings()
+  )(
+    ms: Stream[Throwable, String]
+  ): ZStream[Clock, Throwable, ErrorOrEvent] =
+    eventStream(client, queueUrl, settings)(ms.map(StringEvent))
+
+  def eventStream(
+    client: SqsAsyncClient,
+    queueUrl: String,
+    settings: SqsPublisherStreamSettings = SqsPublisherStreamSettings()
+  )(
+    ms: Stream[Throwable, Event]
+  ): ZStream[Clock, Throwable, ErrorOrEvent] =
+    ms.aggregateAsyncWithin(Sink.collectAllN[Event](settings.batchSize.toLong), Schedule.spaced(settings.duration))
+      .map(buildSendMessageBatchRequest(queueUrl, _))
+      .mapMParUnordered(settings.parallelism)(it => runSendMessageBatchRequest(client, it._1, it._2))
+      .mapConcat(identity)
+
+  private def buildSendMessageBatchRequest(
+    queueUrl: String,
+    ms: List[Event]
+  ): (SendMessageBatchRequest, List[(Event, Int)]) = {
+    val indexedMessages = ms.zipWithIndex
+    val entries = indexedMessages.map {
+      case (m: Event, id: Int) =>
+        SendMessageBatchRequestEntry
+          .builder()
+          .id(id.toString)
+          .messageBody(m.body)
+          .messageAttributes(m.attributes.asJava)
+          .messageGroupId(m.groupId.orNull)
+          .messageDeduplicationId(m.deduplicationId.orNull)
+          .build()
+    }
+
+    val req = SendMessageBatchRequest
+      .builder()
+      .queueUrl(queueUrl)
+      .entries(entries.asJava)
+      .build()
+
+    (req, indexedMessages)
+  }
+
+  def runSendMessageBatchRequest(
+    client: SqsAsyncClient,
+    req: SendMessageBatchRequest,
+    indexedMessages: List[(Event, Int)]
+  ): Task[List[ErrorOrEvent]] =
+    Task.effectAsync[List[ErrorOrEvent]]({ cb =>
+      client
+        .sendMessageBatch(req)
+        .handleAsync[Unit](new BiFunction[SendMessageBatchResponse, Throwable, Unit] {
+          override def apply(res: SendMessageBatchResponse, err: Throwable): Unit =
+            err match {
+              case null =>
+                val m  = indexedMessages.map(it => (it._2.toString, it._1)).toMap
+                val ss = res.successful().asScala.map(it => m(it.id())).map(Right(_): ErrorOrEvent).toList
+                val es = res.failed().asScala.map(err => (err, m(err.id()))).map(Left(_): ErrorOrEvent).toList
+                cb(IO.succeed(ss ++ es))
```

Comment:
@ghostdogpr 
a question:
here we're returning a list of either values -- it means the user of the library should handle failures.
Do we think we need an option in settings to fail the stream here if there is one failed message?

I mean something, like:
```
if(res.failed().nonEmpty)
  cb(IO.fail(....))
```

Reply:
Yeah I think that would be nice! That was what I expected initially even though I can understand the need to receive the full information. Good to have both options.